trieste/models/optimizer.py

```diff
@@ -0,0 +1,118 @@
+from typing import Any, Dict, Optional, Callable, Union, Iterable, Tuple
+from functools import singledispatch
+from dataclasses import dataclass, field
+import tensorflow as tf
+import gpflow
+
+from gpflow.models import ExternalDataTrainingLossMixin, InternalDataTrainingLossMixin
+from ..utils import jit
+from ..datasets import Dataset
+
+
+DatasetBuilder = Callable[[Dataset, Optional[int]], Union[Iterable, Tuple]]
+"""
+Type alias for a function that creates minibatches from a :class:`~optiflow.datasets.Dataset`.
+"""
+
+
+@dataclass
+class Optimizer:
+    optimizer: Union[gpflow.optimizers.Scipy, tf.optimizers.Optimizer]
+    minimize_args: Dict[str, Any] = field(default_factory=lambda: {})
+    compile: bool = False
+
+    def create_loss(self, model: tf.Module, dataset: Dataset) -> Callable:
+        x = tf.convert_to_tensor(dataset.query_points)
+        y = tf.convert_to_tensor(dataset.observations)
+        data = (x, y)
+        return create_loss_function(model, data, self.compile)
+
+    def optimize(self, model: tf.Module, dataset: Dataset):
+        loss_fn = self.create_loss(model, dataset)
+        variables = model.trainable_variables
+        return self.optimizer.minimize(loss_fn, variables, **self.minimize_args)
+
+
+@dataclass
+class TFOptimizer(Optimizer):
```

Comment:
it looks like this class is more about minibatching than tf per se. Is that right? If so, how about `MiniBatchOptimizer`?

Reply:
TF name distinguishes the class of optimizers which are available, not the batching mechanism. Strong _no_ to `MiniBatchOptimizer` :)